File: MATLAB code/dftf.m
%% dftf: 
%   calculates the discrete Fourier transform (DFT) using its real and
%   orthonormal eigenvectors.

% 7/17/2023 FAMU-FSU colledge of engineering
% This file is coded by Rajesh Thomas and modified by Weiwei Wang (ww20br@fsu.edu).
% Supervised by Prof. Victor DeBrunner (victor.debrunner@eng.famu.fsu.edu).
% Reviewed by Prof. Linda DeBrunner (linda.debrunner@eng.famu.fsu.edu).

% Syntax:
%   y = dftf(x)

% Description:
%   y = dftf(x) returns the DFT of the column vector x. The output y is
%   normalized such that the DFT is unitary, i.e.,
%   y = fft(x)/sqrt(length(x)).

% Input Arguments:
%   x: Input signal in a column vector.

% Output Arguments:
%   y: the normalized DFT of x.

%   The eigenvectors are calculated via the dfteig() function which uses
%   Matveev's method to get the orthonormal eigenvectors.

% Example:
% x = [-2; 0; 3; 1; 1; 3; 2];
% y = dftf(x)
% output:
% y =
%   8.0000 + 0.0000i
%  -3.8901 + 1.5637i
%  -6.6039 + 1.9499i
%  -0.5060 + 0.8678i
%  -0.5060 - 0.8678i
%  -6.6039 - 1.9499i
%  -3.8901 - 1.5637i

function y = dftf(x)

if ~iscolumn(x)
    error('Input should be a column vector.');
end
n = size(x,1);
[v, m] = dfteig(n);

idx = cumsum(m);

% combiner matrix
u = zeros(n,2); 
u(1:idx(2),1) = [ones(m(1),1); -ones(m(2),1)];
u(idx(2)+1:end,2) = [ones(m(3),1); -ones(m(4),1)];

xtilde = diag(v.'*x); % output after filtering and downsampling
v_scale = v*xtilde;
y = v_scale*u;
y = y*[1 1j].';
y = sqrt(n)*y;

end

File: Python code/dftf.py
'''
dftf

7/17/2023 FAMU-FSU colledge of engineering
Coded by Weiwei Wang(ww20br@fsu.edu). 
Supervised by Prof. Victor DeBrunner (victor.debrunner@eng.famu.fsu.edu). 
Reviewed by Prof. Linda DeBrunner (linda.debrunner@eng.famu.fsu.edu).

dftf: 
   calculates the discrete Fourier transform (DFT) using its real and orthonormal eigenvectors.

Docstring:
   dftf(x) returns the DFT of the column vector x. The output y is normalized such that the DFT is unitary, i.e., y = fft(x)/sqrt(length(x)).

Parameters:
   x: array_like
      Input signal x.

returns:
   y: array_like
      the normalized DFT of x.

   The eigenvectors are calculated via the dfteig() function which uses Matveev's method to get the orthonormal eigenvectors.

Call signature:
    y = dftf(x)
'''

import torch
import numpy as np
import scipy as sp

from dfteig import dfteig

def dftf(x):
    
    #if not np.ndim(x) == 1:
    #    raise ValueError('Input should be a column vector.')
        
    n = len(x)
    v, m = dfteig(n)
   
    idx = np.cumsum(m)
    idx1 = int(idx[1])
    
    # combiner matrix
    u = np.zeros((n, 2))
    u[:idx1, 0] = np.concatenate((np.ones(int(m[0])), -np.ones(int(m[1]))))
    u[idx1:, 1] = np.concatenate((np.ones(int(m[2])), -np.ones(int(m[3]))))
    
    xtilde = np.diagflat( v.T @ x) # output after filtering and downsampling
    v_scale = v @ xtilde
    y = v_scale @ u                     
    y = y @ np.array([1, 1j])  
    y = np.sqrt(n) * y

    return y